Update Kubernetes Access docs (#5865)
gravitational/teleport: docs/pages/kubernetes-access/getting-started.mdx

description: Getting started with Teleport Kubernetes Access
---

# Getting Started

Let's deploy Teleport in a Kubernetes with SSO and Audit logs:

1. Install Teleport in a Kubernetes cluster.
2. Set up Single Sign On
3. Capture and play back Kubernetes commands.

## Prerequisites

* [Kubernetes](https://kubernetes.io) >= v1.14.0
* [Helm](https://helm.sh) >= v3.2.0

Verify that helm and kubernetes are installed and up to date.

```bash
$ helm version
version.BuildInfo{Version:"v3.4.2"}

$ kubectl version
Client Version: version.Info{Major:"1", Minor:"17+"}
Server Version: version.Info{Major:"1", Minor:"17+"}
```

## Step 1/3 Install Teleport

Let's start with a single-pod Teleport using persistent volume as a backend.

<Tabs>
<TabItem label="Open Source">
```bash
$ helm repo add teleport https://charts.releases.teleport.dev

# Install a single node teleport cluster and provision a cert using ACME.
# Set clusterName to unique hostname, for example teleport.example.com
# Set acmeEmail to receive correspondence from Letsencrypt certificate authority.
$ helm install teleport-cluster teleport-cluster --create-namespace --namespace=teleport-cluster \

>   --set clusterName=${CLUSTER_NAME?} --set acme=true --set acmeEmail=${EMAIL?}

Author: yes, it's in progress in https://github.com/gravitational/next/issues/84

$ gcloud dns record-sets transaction execute --zone="${MYZONE?}"
```
</TabItem>
<TabItem label="AWS Route 53">
```bash
# Tip for finding AWS zone id by the domain name.
$ MYZONE_DNS="example.com"
$ MYZONE=$(aws route53 list-hosted-zones-by-name --dns-name=${MYZONE_DNS?} | jq -r '.HostedZones[0].Id' | sed s_/hostedzone/__)

$ MYDNS="tele.example.com"

# Create a JSON file changeset for AWS.
$ jq -n --arg ip ${MYIP?} --arg dns ${MYDNS?} '{"Comment": "Create records", "Changes": [
  {"Action": "CREATE", "ResourceRecordSet": {"Name": $dns, "Type": "A", "TTL": 300, "ResourceRecords": [{ "Value": $ip}]}},
  {"Action": "CREATE", "ResourceRecordSet": {"Name": ("*." + $dns), "Type": "A", "TTL": 300, "ResourceRecords": [{ "Value": $ip}]}}
  ]}' > myrecords.json

# Review records before applying.
$ cat myrecords.json | jq
# Apply the records and capture change id
$ CHANGEID=$(aws route53 change-resource-record-sets --hosted-zone-id ${MYZONE?} --change-batch file://myrecords.json | jq -r '.ChangeInfo.Id')

# Verify that change has been applied
$ aws route53 get-change --id ${CHANGEID?} | jq '.ChangeInfo.Status'
"INSYNC"
```
</TabItem>

> </Tabs>

Reviewer: In future maybe we could describe how to set up [`external-dns`](https://github.com/bitnami/charts/tree/master/bitnami/external-dns) to make this part automatic. I'm keen not to try and start teaching people how to use Kubernetes/Helm, though.

Author: could be, yes, I saw DNS automation in K8s, but it was quite complex for a getting started guide

The first request to Teleport's API will take a bit longer because it gets
a cert from [Letsencrypt](https://letsencrypt.org).
Teleport will respond back with a discovery info:

```bash
$ curl https://tele.example.com/webapi/ping

{"server_version":"5.0.0-dev","min_client_version":"3.0.0"}
```

## Step 2/3 Create a local admin

Local users are reliable fallback for cases when SSO provider is down.
Let's create a local admin `alice` who has access to Kubernetes group `system:masters`.

Save this role as `member.yaml`:

```yaml
kind: role
version: v3
metadata:
  name: member
spec:
  allow:

>     # This field is used for SSH logins. You have to keep logins non-empty random value

Reviewer: ```suggestion
    # This field is used for SSH logins. You have to keep 'logins' as a non-empty random value
```

# Generate an invite link for the user.
$ kubectl exec -ti ${POD?} tctl -- users add alice --roles=member

User "alice" has been created but requires a password. Share this URL with the user to
complete user setup, link is valid for 1h:

https://tele.example.com:443/web/invite/random-token-id-goes-here

NOTE: Make sure tele.example.com:443 points at a Teleport proxy which users can access.
```

Let's install `tsh` and `tctl` on Linux.
For other install options, check out [install guide](../installation.mdx)

<Tabs>
<TabItem label="Open Source">
```bash
$ curl -L -O https://get.gravitational.com/teleport-v{{ teleport.version }}-linux-amd64-bin.tar.gz
$ tar -xzf teleport-v{{ teleport.version }}-linux-amd64-bin.tar.gz
$ sudo mv teleport/tsh /usr/local/bin/tsh
$ sudo mv teleport/tctl /usr/local/bin/tctl
```
</TabItem>
<TabItem label="Enterprise">
```bash
$ curl -L -O https://get.gravitational.com/teleport-ent-v{{ teleport.version }}-linux-amd64-bin.tar.gz
$ tar -xzf teleport-ent-v{{ teleport.version }}-linux-amd64-bin.tar.gz
$ sudo mv teleport/tsh /usr/local/bin/tsh
$ sudo mv teleport/tctl /usr/local/bin/tctl
```
</TabItem>
</Tabs>

> Try `tsh login` with a local user. Use a custom `KUBECONFIG` to prevent override
> of the default one in case if there is a problem.

Reviewer: ```suggestion
Try `tsh login` with a local user. Use a custom `KUBECONFIG` to prevent overwriting
the default one in case there is a problem.
```

```bash
$ KUBECONFIG=${HOME?}/teleport.yaml tsh login --proxy=tele.example.com:443 --user=alice
```

Teleport updated `KUBECONFIG` with a short-lived 12 hour certificate.

```bash
$ tsh kube ls

Kube Cluster Name Selected
----------------- --------
tele.example.com  *

> # Once working, remove the KUBECONFIG= override to switch to teleport

Reviewer: Not sure what we're trying to explain to people with this line?

Author: It was easy for me to corrupt my kubeconfig when testing this, so wanted to give a tip to users

Reviewer: Yeah, I just didn't quite understand how that would involve switching to Teleport when we'd already overridden the `KUBECONFIG`. I think it's probably clear enough.

NAME                                READY   STATUS    RESTARTS   AGE
teleport-cluster-6c9b88fd8f-glmhf   1/1     Running   0          127m
```

## Step 3/3 SSO for Kubernetes

We are going to setup Github connector for OSS and Okta for Enterpise version.

<Tabs>
<TabItem label="Open Source">

Save the file below as `github.yaml` and update the fields. You will need to set up
[Github OAuth 2.0 Connector](https://developer.github.com/apps/building-oauth-apps/creating-an-oauth-app/) app.
Any member with the team `admin` in the organization `octocats` will be able to assume a builtin role `admin`.
 
```yaml
kind: github
version: v3
metadata:
  # connector name that will be used with `tsh --auth=github login`
  name: github
spec:
  # client ID of Github OAuth app
  client_id: client-id
  # client secret of Github OAuth app
  client_secret: client-secret
  # This name will be shown on UI login screen
  display: Github
  # Change tele.example.com to your domain name
  redirect_url: https://tele.example.com:443/v1/webapi/github/callback
  # Map github teams to teleport roles
  teams_to_logins:
    - organization: octocats # Github organization name

>       team: admin           # Github team name within that organization

Reviewer: ```suggestion
      team: admin            # Github team name within that organization
```

```yaml
kind: saml
version: v2
metadata:
  name: okta
spec:
  acs: https://tele.example.com/v1/webapi/saml/acs
  attributes_to_roles:
  - {name: "groups", value: "okta-admin", roles: ["admin"]}
  entity_descriptor: |
    <?xml !!! Make sure to shift all lines in XML descriptor 
    with 4 spaces, otherwise things will not work
```
</TabItem>
</Tabs>

To create a connector, we are going to run Teleport's admin tool `tctl` from the pod.

<Tabs>
<TabItem label="Open Source">
```bash
# To create a Github connector, we are going to run Teleport's admin tool tctl from the pod.
$ POD=$(kubectl get po -l app=teleport-cluster -o jsonpath='{.items[0].metadata.name}')

$ kubectl exec -i ${POD?} tctl -- create -f < github.yaml
authentication connector "github" has been created
```
</TabItem>
<TabItem label="Enterprise">
```bash
# To create an Okta connector, we are going to run Teleport's admin tool tctl from the pod.
$ POD=$(kubectl get po -l app=teleport-cluster-ent -o jsonpath='{.items[0].metadata.name}')

$ kubectl exec -i ${POD?} tctl -- create -f < okta.yaml
authentication connector 'okta' has been created
```
</TabItem>
</Tabs>

> Try `tsh login` with Github user. I am using a custom `KUBECONFIG` to prevent override
> of the default one in case if there is a problem on the first try.

Reviewer: ```suggestion
Try `tsh login` with Github user. I am using a custom `KUBECONFIG` to prevent overwriting
the default one in case there is a problem.
```

<Tabs>
<TabItem label="Open Source">
```bash
$ KUBECONFIG=${HOME?}/teleport.yaml tsh login --proxy=tele.example.com:443 --auth=github
```
</TabItem>
<TabItem label="Enterprise">
```bash
$ KUBECONFIG=${HOME?}/teleport.yaml tsh login --proxy=tele.example.com:443 --auth=okta
```
</TabItem>
</Tabs>

<Admonition type="note" title="Debugging SSO">

> If you are getting the login error, take a look at the audit log for details:

Reviewer: ```suggestion
If you are getting a login error, take a look at the audit log for details:
```
